ath79: move RE450 V1 & V2 to tiny, restore sysupgrade support

ivanddiaz opened the pull request:
Since OpenWrt 23.05, the firmware selector no longer provides sysupgrade images for RE450 V1 & V2 due to image size constraints. The maximum allowed image size is approximately 5.7MB, making the generic build unusable.

Due to flash size limitations, community members often prioritize LuCI over security, replacing `wpad `with `wpad-mini` and sacrificing support for `wpad-wolfssl` and `wpad-mbedtls`. Without an official build that fits within the flash size limit, users are forced to create custom images, which may lack proper security defaults. Moving RE450 V1 & V2 to the tiny target ensures a secure-by-default wireless setup while allowing full management via UCI commands.

This commit moves RE450 V1 & V2 + RE355 V1 to the tiny target, enabling:
- Smaller image builds within the flash size limit (~5.7MB).
- Better control over package selection while maintaining security defaults.

Image size calculations (custom builds):
```
5.0M openwrt-ath79-tiny-tplink_re355-v1-initramfs-kernel.bin
5.5M openwrt-ath79-tiny-tplink_re355-v1-squashfs-factory.bin
5.5M openwrt-ath79-tiny-tplink_re355-v1-squashfs-sysupgrade.bin

5.0M openwrt-ath79-tiny-tplink_re450-v1-initramfs-kernel.bin
5.5M openwrt-ath79-tiny-tplink_re450-v1-squashfs-factory.bin
5.5M openwrt-ath79-tiny-tplink_re450-v1-squashfs-sysupgrade.bin

5.0M openwrt-ath79-tiny-tplink_re450-v2-initramfs-kernel.bin
5.5M openwrt-ath79-tiny-tplink_re450-v2-squashfs-factory.bin
5.5M openwrt-ath79-tiny-tplink_re450-v2-squashfs-sysupgrade.bin
```

Tested on RE450 V2. Patches and further support available as needed.

robimarko commented on target/linux/ath79/image/tiny-tp-link.mk:
```diff
@@ -1,5 +1,30 @@
+  DEVICE_VENDOR := TP-Link
+  TPLINK_BOARD_ID := RE450
+  SUPPORTED_DEVICES += re450
+  DEFAULT := n
```
I assume you want to drop this to enable it again by default?

ivanddiaz replied:
> I assume you want to drop this to enable it again by default?

Hello @robimarko, right assumption. I've dropped that. :)
Thank you!

robimarko commented:
No merge commits please

Djfe commented:
thanks for taking a look at this :)
Good solution to move the device to tiny. (for the foreseeable future the device has to move there anyways even with my idea below)

I bet, we can also increase available space for the RE450 v1 by combining the `firmware` with the `config` partition using mtd-concat
https://github.com/openwrt/openwrt/blob/main/target/linux/ath79/dts/qca9558_tplink_rex5x.dtsi#L163
This will increase the partition size from 5.875MiB to 7.625MiB.

The same is possible for the RE450 v2 as well.
https://github.com/openwrt/openwrt/blob/main/target/linux/ath79/dts/qca9563_tplink_re450-v2.dts#L53
The partition layout differs a bit from v1. Here config is 0x30000 large instead of 0x1c0000, but there is no partition between config and art.
If config were to fully use the available space it's 0x1c0000 again, leading to the same 7.625MiB of available space.

The only thing I'm not certain about: what does TP-Link save in "config" => is it just temporary data and not important?
I bought an RE450 v1 two months ago and could take a look. Until then, feel free to merge this PR :)
I have no clue when I'll get to work on this

ivanddiaz commented:
> No merge commits please

@robimarko My bad, I've reverted this.

Thank you!

ivanddiaz commented:
> thanks for taking a look at this :) (...) have no clue when I'll get to work on this

@Djfe Thanks for your comments. I own a V2 so whenever you can I can collaborate doing tests.

piaddic7ed commented:
If you need help testing let me know. I have a bunch of v1, v1.1 and v2 devices here.

hauke commented on target/linux/ath79/image/tiny-tp-link.mk:
```diff
@@ -1,5 +1,29 @@
 include ./common-tp-link.mk
 
+define Device/tplink_re450-v1
+  $(Device/tplink_rex5x-v1)
+  SOC := qca9558
+  IMAGE_SIZE := 6016k
+  DEVICE_MODEL := RE450
+  DEVICE_VARIANT := v1
+  DEVICE_VENDOR := TP-Link
+  TPLINK_BOARD_ID := RE450
+  SUPPORTED_DEVICES += re450
```
`Device/tplink_rex5x-v1` is not defined in this file.

You should move `define Device/tplink_rex5x-v1` to `target/linux/ath79/image/common-tp-link.mk`. Then you can just copy the definition from the `generic-tp-link.mk` file here.

ivanddiaz replied:
This is solved in https://github.com/openwrt/openwrt/pull/18126/commits/f6ad2a95bfc9d3b457e6c168153a20de02c9b874

hauke commented:
Please use the first commit to just copy the definition to the tiny subtarget and then a separate commit to remove the `DEFAULT := n` to build it.

Commit:
ath79: move TP-Link RE450 v1 & v2 and RE355 v1 to tiny target

This restores sysupgrade support

Since OpenWrt 23.05, the firmware selector no longer provides sysupgrade
images for RE450 V1 & V2 and RE355 v1 due to image size constraints
(~5.7MB limit), making the generic build unusable.

This commit moves RE450 V1 & V2 to the tiny target, enabling:
- Smaller image builds within the flash size limit.
- Restored sysupgrade support via the firmware selector.
- A secure-by-default wireless setup while allowing full management via UCI.
- Officially supported images, reducing the need for custom (potentially insecure) builds.

This ensures users can install a functional OpenWrt build without compromising security defaults.

Signed-off-by: Ivan Diaz <diaz.it@icloud.com>
[Squashed the commits together and just copied the target definition
from the generic subtarget.]
Link: https://github.com/openwrt/openwrt/pull/18126
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ivanddiaz commented:
> Please use the first commit to just copy the definition to the tiny subtarget and then a separate commit to remove the `DEFAULT := n` to build it.

Hi @hauke ,

I’d like to apologize for not fully understanding the process of tracking approvers' feedback earlier, and I really appreciate your guidance on this. Thank you!

I've now made the two commits as mentioned and also added a separate commit to align with the latest feedback received.

ivanddiaz commented:
> If you need help testing let me know. I have a bunch of v1, v1.1 and v2 devices here.

Hi @piaddic7ed,
Thanks for your comments! I own a V2 and have been using it with this build variant daily. I’d appreciate it if you could test a V1.

DragonBluep commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -554,28 +546,6 @@ define Device/tplink_re355-v1
-  DEVICE_PACKAGES := kmod-ath10k-ct-smallbuffers ath10k-firmware-qca988x-ct
-  TPLINK_BOARD_ID := RE450-V2
-  LOADER_TYPE := elf
-  DEFAULT := n
```
I believe 6016k is still not enough to generate a tiny image. D-Link DIR-859 has the same default packages with this device and its image size is 6080.3 KB.

ref: https://downloads.openwrt.org/snapshots/targets/ath79/tiny/

ivanddiaz replied:
@DragonBluep, here are my build results for RE450 V1/V2 using tiny:
![imagen](https://github.com/user-attachments/assets/f2356fd7-277c-4e78-bd87-d620a7868983)

These images boot and persist settings correctly.
![imagen](https://github.com/user-attachments/assets/2b3a2a59-f427-4f2a-858a-2ed68b85c150)
![imagen](https://github.com/user-attachments/assets/11f6dc6b-0258-4ff5-b1f0-277f5bd175ee)

I'm using it as home repeater using this build. I'd like to be more efficient posting evidence. I'm unsure how D-Link DIR-859 is built but in TP-Link RE-450 v1/v2 I was able to obtain a smaller image without modifications.

Thanks for reading :).

DragonBluep replied:
The official buildbot will generate larger size images compared to our custom build because it will build all kernel modules.

ivanddiaz replied:
Hello @DragonBluep,

You're right.

This PR might not work as intended, but please consider this: the device won't function as a generic target, even with a slim kernel image. However, a home-built tiny image fits and even leaves room for extra packages if needed.

The issue is that buildbot builds exceed the size limit, but home builds will work if we move it to tiny—requiring no extra effort beyond cloning, selecting, and making. While this might not align with having images in the firmware selector, it still provides an option for those who want to extend the device’s life with minimal effort.

This is a solid SoC, now WPA3-capable, and having an official way to support it—even if only through home builds—could benefit everyone. I’d appreciate it if you reconsider this move to give the community the option to build it themselves.

DragonBluep replied:
> Hello @DragonBluep,
> 
> You're right.
> 
> This PR might not work as intended, but please consider this: the device won't function as a generic target, even with a slim kernel image. However, a home-built tiny image fits and even leaves room for extra packages if needed.

It's fine to move them to tiny target. Just don't remove `DEFAULT := n`. I'm pretty sure the official buildbot can't generate images for them.

> 
> The issue is that buildbot builds exceed the size limit, but home builds will work if we move it to tiny—requiring no extra effort beyond cloning, selecting, and making. While this might not align with having images in the firmware selector, it still provides an option for those who want to extend the device’s life with minimal effort.
> 
> This is a solid SoC, now WPA3-capable, and having an official way to support it—even if only through home builds—could benefit everyone. I’d appreciate it if you reconsider this move to give the community the option to build it themselves.

ivanddiaz commented:
> Can you please move the RE355 as well? It has the same issue and is more or less a clone of the RE450v1 afaik.

Hi @achterin 

As this PR focuses on the RE450 v1/v2, I’m unsure if it’s appropriate to expand the scope to include the RE355. I understand they are similar devices, and I have no issue making the change if the maintainers approve it. If they’re on board, I’d be happy to include it.

ivanddiaz commented:
> I bet, we can also increase available space for the RE450 v1 by combining the `firmware` with the `config` partition using mtd-concat https://github.com/openwrt/openwrt/blob/main/target/linux/ath79/dts/qca9558_tplink_rex5x.dtsi#L163 This will increase the partition size from 5.875MiB to 7.625MiB.

@Djfe took time for me understanding this. That would give for sure a second life to RE's.

ivanddiaz commented:
> Can you please move the RE355 as well? It has the same issue and is more or less a clone of the RE450v1 afaik.

Hi @achterin, this is done https://github.com/openwrt/openwrt/pull/18126/commits/99231acdc837ec51a771590033281bcba64d02e4

hauke commented:
I squashed it together as the other  commits are only adding and removing the `DEFAULT := n` attribute. 

Please add a real email address.

DragonBluep reviewed (changes requested):
Please use `git commit --amend --author="Ivan Diaz <diaz.it@icloud.com>" --no-edit` to fix the author name and email.

DragonBluep commented on target/linux/ath79/image/tiny-tp-link.mk:
```diff
@@ -1,5 +1,44 @@
+  IMAGE_SIZE := 6016k
+  DEVICE_MODEL := RE450
+  DEVICE_VARIANT := v2
+  DEVICE_PACKAGES := kmod-ath10k-ct-smallbuffers ath10k-firmware-qca988x-ct
```
Maybe we can remove ath10k package to get more free spaces? User can manually select them on they custom build. In this way, we can get an official build image with 2 GHz WiFi only. Let's wait other reviewer's comment.

ivanddiaz commented:
Hello, @robimarko @hauke @DragonBluep 

First of all thanks for all the guidance and help doing a simple thing like this.

I understand that must need three additional reviewers, meaning we're in 50%. Is there anything else I can do from my end?
Any recommendation to bump the process? Maybe via IRC or something?. May be is preferred just to wait?

Thanks again!

DragonBluep reviewed (commented):
> I understand that must need three additional reviewers, meaning we're in 50%. Is there anything else I can do from my end?

Please be patient. Maintainers can push it to the mainline even if there are no more reviewers.

masterblaster79 commented:
> I bet, we can also increase available space for the RE450 v1 by combining the `firmware` with the `config` partition using mtd-concat https://github.com/openwrt/openwrt/blob/main/target/linux/ath79/dts/qca9558_tplink_rex5x.dtsi#L163 This will increase the partition size from 5.875MiB to 7.625MiB.

Hey there. I also own an RE450 v1 and would like to express my interest in this. It would be just great to get a recent tiny image from downloads.openwrt.org for this device.
Would it help if I take a dump of the config partition and attach it here?

ivanddiaz commented:
Hello,

First of all, thanks for all the help, support and patience reading my PR and waiting for the corrections. I’ve learned a lot doing this and I want to keep committed to contributing to this amazing project.

I know @achterin and @masterblaster79 have talked before about the possibility to expand the firmware image and bring back RE350/450 v1/v2 to the firmware selector. This is a good piece of hardware, small and fast.

Now that the PR work is finished, I’d like to invite you to continue the discussion in the right place. I made a forum post talking 
about this PR and the lessons I learned:
[forum.openwrt.org> TP-Link RE355 v1 and RE450 v1/v2 moved to tiny builds](https://forum.openwrt.org/t/tp-link-re355-v1-and-re450-v1-v2-moved-to-tiny-builds/241895)

If you want to follow the conversation or share suggestions, you are very welcome.
Thanks again!
